api/src/main/java/org/openmrs/module/fhirExtension/service/ObsBasedDiagnosticReportService.java

```diff
@@ -96,11 +148,121 @@ public DiagnosticReport create(@Nonnull DiagnosticReport diagnosticReport) {
+	
+	private Encounter createNewEncounterForReport(FhirDiagnosticReport fhirDiagnosticReport, Order order) {
+		if (fhirDiagnosticReport.getEncounter() != null) {
+			log.info("Diagnostic Report was submitted with an existing encounter reference. This will be overwritten by a new encounter");
+		}
+		
+		EncounterType encounterType = encounterService.getEncounterType(LAB_RESULT_ENC_TYPE);
+		if (encounterType == null) {
+			log.error("Encounter type LAB_RESULT must be defined to support Diagnostic Report");
+			throw new RuntimeException(UNABLE_TO_PROCESS_DIAGNOSTIC_REPORT);
+		}
+		
+		Location location = Context.getUserContext().getLocation(); //TODO if not present get from clinic
+		if (location == null) {
+			log.error("Logged in location for user is null. Can not identify encounter session.");
+			throw new RuntimeException(UNABLE_TO_PROCESS_DIAGNOSTIC_REPORT);
+		}
+
+		Optional<Visit> reportOrderVisit = Optional.ofNullable(order).map(ord -> ord.getEncounter()).map(enc -> enc.getVisit());
+		Visit applicableVisit = reportOrderVisit.isPresent() ? reportOrderVisit.get() : getActiveVisit(fhirDiagnosticReport).orElseGet(() -> {
+			log.warn("Can not identify an active visit for the patient. Trying to identify a lab visit for today...");
+			return findOrCreateLabVisit(fhirDiagnosticReport, location);
+		});
+		if (applicableVisit == null) {
+			log.error("Can not identify or create visit for the patient for lab results upload. Please check with your administrator");
+			throw new RuntimeException(UNABLE_TO_PROCESS_DIAGNOSTIC_REPORT);
+		}
+		
+		return encounterService.saveEncounter(newEncounterInstance(fhirDiagnosticReport.getSubject(), encounterType,
+		    location, applicableVisit, Context.getAuthenticatedUser()));
+	}
+	
+	private Visit findOrCreateLabVisit(FhirDiagnosticReport report, Location location) {
+		String labEntryVisitType = adminService.getGlobalProperty(LAB_ENTRY_VISIT_TYPE);
+		if (labEntryVisitType == null || "".equals(labEntryVisitType)) {
+			labEntryVisitType = DEFAULT_LAB_VISIT_TYPE;
+		}
+		List<VisitType> labVisitTypes = visitService.getVisitTypes(labEntryVisitType);
+		if (CollectionUtils.isEmpty(labVisitTypes)) {
+			return null;
+		}
+		LocalDate today = LocalDate.now();
+		Instant startOfDay = today.atStartOfDay().atZone(ZoneId.systemDefault()).toInstant();
+		Instant endOfDate = LocalTime.MAX.atDate(today).atZone(ZoneId.systemDefault()).toInstant();
+		Date startDate = Date.from(startOfDay);
+		Date endDate = Date.from(endOfDate);
+
+		List<Visit> labVisitsForToday = visitService.getVisits(
+				labVisitTypes,
+				Collections.singletonList(report.getSubject()),
+				null,
+				null,
+				startDate, null,
+				null, endDate,
+				null,
+				true,
+				false);
+
+		if (!CollectionUtils.isEmpty(labVisitsForToday)) {
+			return labVisitsForToday.stream().reduce((first, second) -> second).orElse(null);
```

Comment:
@angshu here, we are returning the earliest visit of the day where as in [getActiveVisit()](https://github.com/Bahmni/openmrs-module-fhir2Extension/pull/17/files#diff-2f8861b0efed9655fd54f6939486601a9b3653b6050a6f165241d36baef2789bR253) returning latest visit of the day.

Reply:
do we need to update the `endDate` for selected inactive visit ?

Reply:
We are trying to find out in case we are going to create a LAB_VISIT, if there is already one for the day. If found, we associate the results against that.